src/mbgl/text/placement.cpp

```diff
@@ -1186,50 +1240,88 @@ void StaticPlacement::commit() {
+    void placeSymbolBucket(const BucketPlacementData&, std::set<uint32_t>&) override;
     void collectPlacedSymbolData(bool enable) override { collectData = enable; }
     const std::vector<PlacedSymbolData>& getPlacedSymbolsData() const override { return placedSymbolsData; }
+
     optional<CollisionBoundaries> getAvoidEdges(const SymbolBucket&, const mat4&) override;
-    SymbolInstanceReferences getSortedSymbols(const BucketPlacementData&, float pixelRatio) override;
-    bool stickToFirstVariableAnchor(const CollisionBox& box,
-                                    Point<float> shift,
-                                    const mat4& posMatrix,
-                                    float textPixelRatio) override;
+    bool canPlaceAtVariableAnchor(const CollisionBox& box,
+                                  TextVariableAnchorType anchor,
+                                  Point<float> shift,
+                                  std::vector<style::TextVariableAnchorType>& anchors,
+                                  const mat4& posMatrix,
+                                  float textPixelRatio) override;
     void newSymbolPlaced(const SymbolInstance&,
                          const JointPlacement&,
                          style::SymbolPlacementType,
                          const std::vector<ProjectedCollisionBox>&,
                          const std::vector<ProjectedCollisionBox>&) override;
+    void populateIntersectingSymbols();
 
     std::unordered_map<uint32_t, bool> locationCache;
     optional<CollisionBoundaries> tileBorders;
     std::set<uint32_t> seenCrossTileIDs;
     std::vector<PlacedSymbolData> placedSymbolsData;
-    bool onlyLabelsIntersectingTileBorders = false;
+    std::vector<Intersection> intersections;
+    bool populateIntersections = false;
     bool collectData = false;
 };
 
 void TilePlacement::placeLayers(const RenderLayerReferences& layers) {
-    // In order to avoid label cut-offs, at first, place the labels,
-    // which cross tile boundaries.
-    onlyLabelsIntersectingTileBorders = true;
     placedSymbolsData.clear();
     seenCrossTileIDs.clear();
+    intersections.clear();
+    // Populale intersections.
+    populateIntersections = true;
     for (auto it = layers.crbegin(); it != layers.crend(); ++it) {
         placeLayer(*it, seenCrossTileIDs);
     }
+
+    std::stable_sort(intersections.begin(), intersections.end(), [](const Intersection& a, const Intersection& b) {
+        uint8_t flagsA = a.status.flags;
+        uint8_t flagsB = b.status.flags;
+        // Items arranged as: VerticalBorders & HorizontalBorders (3) ->
+        // VerticalBorders (2) -> HorizontalBorders (1)
+        if (flagsA != flagsB) return flagsA > flagsB;
+        // If both intersects the same border(s), look for a more noticable cut-off.
+        if (a.status.minSectionLength != b.status.minSectionLength) {
+            return a.status.minSectionLength > b.status.minSectionLength;
```

Comment:
Should vertical and horizontal length be checked separately? Also, should it be total overlap area vs single 'section length'? Label that is in a tile corner can have bigger overlap area, yet, smaller section length.

Reply:
Thanks for the proposal! will consider it separately